File: frontend/myNewApp/src/services/userProfileService.js
import axios from 'axios';
import authService from './authService';

const API_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8080';

// Create an axios instance with base configuration
const api = axios.create({
  baseURL: API_URL,
  headers: {
    'Content-Type': 'application/json'
  }
});

// Add a request interceptor to include the JWT token in all authenticated requests
api.interceptors.request.use(
  (config) => {
    const token = authService.getToken();
    if (token) {
      config.headers['Authorization'] = `Bearer ${token}`;
    }
    return config;
  },
  (error) => {
    return Promise.reject(error);
  }
);

// Log requests in development - helpful for debugging
if (import.meta.env.DEV) {
  api.interceptors.request.use(request => {
    console.log('Profile API Request:', request.method?.toUpperCase(), request.baseURL + request.url);
    return request;
  });
  
  api.interceptors.response.use(
    response => {
      console.log('Profile API Response:', response.status, response.config.url);
      return response;
    },
    error => {
      console.error('Profile API Error:', error.response?.status, error.config?.url, error.message);
      return Promise.reject(error);
    }
  );
}

// User Profile service
const userProfileService = {
  // Get user profile
  getUserProfile: async () => {
    try {
      const response = await api.get('/user/profile');
      return response.data;
    } catch (error) {
      console.error('Error fetching user profile:', error.response?.data || error.message);
      throw error.response?.data || { error: error.message };
    }
  },
  
  // Create or update user profile
  updateUserProfile: async (profileData) => {
    try {
      const response = await api.post('/user/profile', profileData);
      return response.data;
    } catch (error) {
      console.error('Error updating user profile:', error.response?.data || error.message);
      throw error.response?.data || { error: error.message };
    }
  },
  
  // Helper function to format profile data for API
  formatProfileData: (formData) => {
    const { 
      firstName, 
      lastName, 
      dateOfBirth, 
      gender, 
      bloodGroup, 
      height, 
      weight, 
      phone, 
      address, 
      emergencyContact,
      allergies,
      medicalConditions,
      medications 
    } = formData;
    
    // Convert string fields for allergies/conditions/medications to arrays if needed
    const formatStringToArray = (value) => {
      if (!value) return [];
      if (Array.isArray(value)) return value;
      return value.split(',').map(item => item.trim()).filter(item => item);
    };
    
    return {
      firstName,
      lastName,
      dateOfBirth,
      gender,
      bloodGroup,
      height: height ? parseFloat(height) : null,
      weight: weight ? parseFloat(weight) : null,
      phone,
      address,
      emergencyContact,
      allergies: formatStringToArray(allergies),
      medicalConditions: formatStringToArray(medicalConditions),
      medications: formatStringToArray(medications)
    };
  }
};

export default userProfileService; 

File: backend/src/main/kotlin/com/jeevan/backend/services/UserProfileService.kt
package com.jeevan.backend.services

import com.jeevan.backend.dto.UserProfileRequest
import com.jeevan.backend.dto.UserProfileResponse
import com.jeevan.backend.models.User
import com.jeevan.backend.models.UserProfile
import com.jeevan.backend.repositories.UserProfileRepository
import org.springframework.stereotype.Service
import java.time.LocalDateTime

@Service
class UserProfileService(
    private val userProfileRepository: UserProfileRepository
) {
    fun getUserProfile(user: User): UserProfileResponse {
        val profileOptional = userProfileRepository.findByUser(user)
        
        return if (profileOptional.isPresent) {
            val profile = profileOptional.get()
            mapToResponse(profile)
        } else {
            UserProfileResponse(userId = user.id, message = "Profile not found")
        }
    }
    
    fun createOrUpdateProfile(user: User, request: UserProfileRequest): UserProfileResponse {
        val existingProfile = userProfileRepository.findByUser(user)
        
        val profile = if (existingProfile.isPresent) {
            // Update existing profile
            val current = existingProfile.get()
            val updated = current.copy(
                firstName = request.firstName ?: current.firstName,
                lastName = request.lastName ?: current.lastName,
                dateOfBirth = request.dateOfBirth ?: current.dateOfBirth,
                gender = request.gender ?: current.gender,
                bloodGroup = request.bloodGroup ?: current.bloodGroup,
                height = request.height ?: current.height,
                weight = request.weight ?: current.weight,
                phone = request.phone ?: current.phone,
                address = request.address ?: current.address,
                emergencyContact = request.emergencyContact ?: current.emergencyContact,
                allergies = request.allergies?.joinToString(",") ?: current.allergies,
                medicalConditions = request.medicalConditions?.joinToString(",") ?: current.medicalConditions,
                medications = request.medications?.joinToString(",") ?: current.medications,
                updatedAt = LocalDateTime.now()
            )
            userProfileRepository.save(updated)
        } else {
            // Create new profile
            val newProfile = UserProfile(
                user = user,
                firstName = request.firstName,
                lastName = request.lastName,
                dateOfBirth = request.dateOfBirth,
                gender = request.gender,
                bloodGroup = request.bloodGroup,
                height = request.height,
                weight = request.weight,
                phone = request.phone,
                address = request.address,
                emergencyContact = request.emergencyContact,
                allergies = request.allergies?.joinToString(","),
                medicalConditions = request.medicalConditions?.joinToString(","),
                medications = request.medications?.joinToString(",")
            )
            userProfileRepository.save(newProfile)
        }
        
        return mapToResponse(profile, "Profile saved successfully")
    }
    
    private fun mapToResponse(profile: UserProfile, message: String? = null): UserProfileResponse {
        return UserProfileResponse(
            id = profile.id,
            userId = profile.user.id,
            firstName = profile.firstName,
            lastName = profile.lastName,
            dateOfBirth = profile.dateOfBirth,
            gender = profile.gender,
            bloodGroup = profile.bloodGroup,
            height = profile.height,
            weight = profile.weight,
            phone = profile.phone,
            address = profile.address,
            emergencyContact = profile.emergencyContact,
            allergies = profile.allergies?.split(",")?.filter { it.isNotBlank() },
            medicalConditions = profile.medicalConditions?.split(",")?.filter { it.isNotBlank() },
            medications = profile.medications?.split(",")?.filter { it.isNotBlank() },
            message = message
        )
    }
} 